Don't revert upcast for persistent schedulers to avoid increasing persistent buffer sizes

liqiangxl opened the pull request:
**Change:**
Don't revert upcast for persistent schedulers to avoid increasing persistent buffer sizes

**Influence:**
Fix test `mpirun -np 1 pytest tests/python/test_multidevice.py --only-mpi -s -k transformer_backward`
Before:
```
T62_l_float[bS202{1}, iS203{2048}, iS204{12288}] (DeviceMesh{0})
   = (float)(T60_g_bool[bS196{1}, iS197{2048}, iS198{12288}] (DeviceMesh{0}));
Use T62_l_float as persistent buffer
```
After:
```
T286_l_float[bS907{1}, iS908{2048}, iS909{12288}] (DeviceMesh{0})
   = (float)(T60_g_bool[bS196{1}, iS197{2048}, iS198{12288}] (DeviceMesh{0}));
T62_l_float[bS202{1}, iS203{2048}, iS204{12288}] (DeviceMesh{0})
   = (float)(T60_g_bool[bS196{1}, iS197{2048}, iS198{12288}] (DeviceMesh{0}));
Use T60_g_bool as persistent buffer
```

liqiangxl commented:
!test

liqiangxl commented:
!test --diff

liqiangxl commented:
!build

liqiangxl commented:
!test

liqiangxl commented:
!test

naoyam commented:
This seems to indicate the real problem is the buffer selection algorithm. What do you think?

liqiangxl commented:
> This seems to indicate the real problem is the buffer selection algorithm. What do you think?

Yes, the current algorithm does't consider recompute an internal tensor from another internal non-input tensor. We can revise to add this capability. If we only consider castOp, should be a straightforward extension. Let me try it in another PR.

The idea for this fix is: The Segmenter finds a scheduler for the fusion; therefore, we should avoid modifying the fusion unless absolutely necessary, as it may invalidate the scheduler.

liqiangxl commented:
> mpirun -np 1 pytest tests/python/test_multidevice.py --only-mpi -s -k transformer_backward



> > The idea for this fix is: The Segmenter finds a scheduler for the fusion; therefore, we should avoid modifying the fusion unless absolutely necessary, as new changes may invalidate the scheduler.
> 
> When I added the cast changes, I tried to make it effectively no-op for schedulers. It should be the case mathematically as using a common cast op or separate cast ops shouldn't matter. If that difference resulted in some different scheduling heuristics, then it seems to me that's an indication of lack of robustness of the scheduler. Namely, in this case, using a common or separate cast ops shouldn't matter for the normalization schedulers, and if not, I'd consider it's a limitation of the scheduler. If that's not easily fixable, I'd be happy to reconsider the cast insertion-reversion approach.

I already added a fix at https://github.com/NVIDIA/Fuser/pull/4051, it only considers `cast op`

liqiangxl commented:
No longer required. Original issue fixed in #4051
